knoten/bundle.py

```diff
@@ -373,3 +406,107 @@ def compute_residuals(network, sensors):
 
     V = V.reshape(num_meas*2)
     return V
+
+def update_parameters(sensors, parameters, network, updates, coefficient_columns):
+    """
+    Updates the sensor objects parameter values and the ground point values in the
+    networks DataFrame. The update occurs directly to variables, so nothing is returned.
+
+    Parameters
+    ----------
+    sensors  :  dict
+                A dictionary that maps ISIS serial numbers to CSM sensors
+    parameters :  list
+                  The list of  CsmParameter to compute the partials W.R.T.
+    network  :  DataFrame
+                The control network as a dataframe generated by plio.
+    updates   : np.ndarray
+                An array of updated parameter values
+    coefficient_columns:  OrderedDict
+                          Dictionary that maps serial numbers and point IDs to
+                          the column range their parameters are in the Jacobian
+                          matrix.
+
+    Returns
+    -------
+
+    """
+    # update the sensor partials
+    for sn, sensor in sensors.items():
+        for i, param in enumerate(parameters[sn]):
+            if i > coefficient_columns[sn][1]:
+                print('THIS SHOULD BE AN ACTUAL ERROR')
+            current_value = sensor.getParameterValue(param.index)
+            sensor.setParameterValue(param.index, current_value+updates[coefficient_columns[sn][0]+i])
+
+    # update ground points
+    for _, row in network.iterrows():
+        point_id = row['id']
+        ground_pt = row[['adjustedX', 'adjustedY', 'adjustedZ']].values
+        adj = updates[coefficient_columns[point_id][0]:coefficient_columns[point_id][1]]
+        network.loc[network.id == point_id, ["adjustedX", "adjustedY", "adjustedZ"]] = ground_pt + adj
+
+def compute_sigma(V, dX, W_parameters, W_observations):
+    """
+    Computes the resulting standard deviation of the residuals for the current state of the bundle network.
+
+    Parameters
+    ----------
+    V : ndarray
+        An array of residuals of the difference between registered measure
+        and back projected ground points in image space.
+    dX : ndarray
+         The array of parameter updates
+    W_parameters: ndarray
+                  The parameter weight matrix (i.e.: sensor parameters and point weights)
+    W_observations : ndarray
+                     The observation weight matrix (i.e.: measure weights)
+
+    Returns
+    -------
+       : float64
+         Standard deviation of the residuals
+
+    """
+    num_parameters = W_parameters.shape[0]
+    num_observations = W_observations.shape[0]
+    dof = num_observations - num_parameters
+    VTPV = V.dot(W_observations).dot(V) + dX.dot(W_parameters).dot(dX)
+    sigma0 = np.sqrt(VTPV/dof)
```

Comment:
Why is it simga0 if it's just the sigma at the current state? This seems consistent with everything else, but I'm still confused by this.

Reply:
sigma0 doesn't have anything to do with iterations. It is technically, the variance of an observation of unit weight. It's a weirdly complex thing. There are multiple sigmas for different observations and parameters, but there is only ever 1 sigma0. I think the names for the functions need to be updated to compute_sigma0.